Firebase/Database/Api/FIRDatabaseComponent.m

```diff
@@ -0,0 +1,219 @@
+/*
+ * Copyright 2018 Google
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#import "FIRDatabaseComponent.h"
+
+#import "FIRDatabase_Private.h"
+#import "FIRDatabaseConfig_Private.h"
+#import "FRepoManager.h"
+
+#import <FirebaseAuthInterop/FIRAuthInterop.h>
+#import <FirebaseCore/FIRAppInternal.h>
+#import <FirebaseCore/FIRComponent.h>
+#import <FirebaseCore/FIRComponentContainer.h>
+#import <FirebaseCore/FIRComponentRegistrant.h>
+#import <FirebaseCore/FIRDependency.h>
+#import <FirebaseCore/FIROptions.h>
+
+NS_ASSUME_NONNULL_BEGIN
+
+/** A NSMutableDictionary of FirebaseApp name and FRepoInfo to FirebaseDatabase instance. */
+typedef NSMutableDictionary<NSString *, NSMutableDictionary<FRepoInfo *, FIRDatabase *> *>
+    FIRDatabaseDictionary;
+
+@interface FIRDatabase ()
+@property (nonatomic, strong) FRepoInfo *repoInfo;
+@property (nonatomic, strong) FIRDatabaseConfig *config;
+@property (nonatomic, strong) FRepo *repo;
+
+- (id)initWithApp:(FIRApp *)app repoInfo:(FRepoInfo *)info config:(FIRDatabaseConfig *)config;
+@end
+
+@interface FIRDatabaseComponent () <FIRComponentRegistrant>
+/// Internal intializer.
+- (instancetype)initWithApp:(FIRApp *)app;
+@end
+
+@implementation FIRDatabaseComponent
+
+#pragma mark - Initialization
+
+- (instancetype)initWithApp:(FIRApp *)app {
+  self = [super init];
+  if (self) {
+    _app = app;
+  }
+  return self;
+}
+
+#pragma mark - Lifecycle
+
++ (void)load {
+  [FIRComponentContainer registerAsComponentRegistrant:self];
+  NSNotificationCenter *center = [NSNotificationCenter defaultCenter];
+  [center addObserverForName:kFIRAppDeleteNotification
```

Comment:
There's a new protocol, `FIRComponentLifecycleMaintainer` that has an `appWillBeDeleted:` call that can replace this. The component container will loop through cached instances when an app is deleted and if the instance conforms to `FIRComponentLifecycleMaintainer` will call `appWillBeDeleted:`.

Reply:
Done